src/types/weakref/proxy.rs

```diff
@@ -0,0 +1,804 @@
+    ///
+    ///     assert_eq!(
+    ///         parse_data(reference.as_borrowed())?,
+    ///         "The object, which this reference refered to, no longer exists"
+    ///     );
+    ///
+    ///     Ok(())
+    /// })
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    /// This function panics is the current object is invalid.
+    /// If used propperly this is never the case. (NonNull and actually a weakref type)
+    ///
+    /// [`PyWeakref_GetObject`]: https://docs.python.org/3/c-api/weakref.html#c.PyWeakref_GetObject
+    /// [`weakref.ProxyType`]: https://docs.python.org/3/library/weakref.html#weakref.ProxyType
+    /// [`weakref.proxy`]: https://docs.python.org/3/library/weakref.html#weakref.proxy
+    pub fn get_object(&self) -> Option<&'_ PyAny> {
+        self.as_borrowed().get_object().map(Bound::into_gil_ref)
+    }
+
+    /// Retrieve to a object pointed to by the weakref.
+    ///
+    /// This function returns `&'py PyAny`, which is either the object if it still exists, otherwise it will refer to [`PyNone`](crate::types::none::PyNone).
+    ///
+    /// This function gets the optional target of this [`weakref.ProxyType`] (result of calling [`weakref.proxy`]).
+    /// It produces similair results using [`PyWeakref_GetObject`] in the C api.
+    ///
+    /// # Example
+    /// ```rust
+    /// use pyo3::prelude::*;
+    /// use pyo3::types::PyWeakProxy;
+    ///
+    /// #[pyclass(weakref)]
+    /// struct Foo { /* fields omitted */ }
+    ///
+    /// fn get_class(reference: Borrowed<'_, '_, PyWeakProxy>) -> PyResult<String> {
+    ///     reference
+    ///         .get_object_raw()
+    ///         .getattr("__class__")?
+    ///         .repr()?
+    ///         .to_str()
+    ///         .map(ToOwned::to_owned)
+    /// }
+    ///
+    /// # fn main() -> PyResult<()> {
+    /// Python::with_gil(|py| {
+    ///     let object = Bound::new(py, Foo{})?;
+    ///     let reference = PyWeakProxy::new_bound(py, object.clone())?;
+    ///
+    ///     assert_eq!(
+    ///         get_class(reference.as_borrowed())?,
+    ///         "<class 'builtins.Foo'>"
+    ///     );
+    ///
+    ///     drop(object);
+    ///
+    ///     assert_eq!(get_class(reference.as_borrowed())?, "<class 'NoneType'>");
+    ///
+    ///     Ok(())
+    /// })
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    /// This function panics is the current object is invalid.
+    /// If used propperly this is never the case. (NonNull and actually a weakref type)
+    ///
+    /// [`PyWeakref_GetObject`]: https://docs.python.org/3/c-api/weakref.html#c.PyWeakref_GetObject
+    /// [`weakref.ProxyType`]: https://docs.python.org/3/library/weakref.html#weakref.ProxyType
+    /// [`weakref.proxy`]: https://docs.python.org/3/library/weakref.html#weakref.proxy
+    pub fn get_object_raw(&self) -> &'_ PyAny {
+        self.as_borrowed().get_object_raw().into_gil_ref()
+    }
+}
+
+impl<'py> PyWeakRefMethods<'py> for Bound<'py, PyWeakProxy> {
```

Comment:
This is the first case of reusing a `Methods` trait for multiple types; I guess it makes sense to do this 👍